src/preprocessing/particle_packing/system.jl

```diff
@@ -0,0 +1,283 @@
+                       update_from_callback=false)
+    if !update_from_callback && !(system.update_callback_used[])
+        throw(ArgumentError("`UpdateCallback` is required when using `ParticlePackingSystem`"))
+    end
+
+    return system
+end
+
+function constrain_particles_onto_surface!(u, system::ParticlePackingSystem)
+    (; neighborhood_search, signed_distance_field) = system
+    (; positions, distances, normals) = signed_distance_field
+
+    search_radius2 = compact_support(system, system)^2
+
+    @threaded system for particle in eachparticle(system)
+        particle_position = current_coords(u, system, particle)
+
+        volume = zero(eltype(system))
+        distance_signed = zero(eltype(system))
+        normal_vector = fill(volume, SVector{ndims(system), eltype(system)})
+
+        # Interpolate signed distances and normals
+        for neighbor in PointNeighbors.eachneighbor(particle_position, neighborhood_search)
+            pos_diff = positions[neighbor] - particle_position
+            distance2 = dot(pos_diff, pos_diff)
+            distance2 > search_radius2 && continue
+
+            distance = sqrt(distance2)
+            kernel_weight = smoothing_kernel(system, distance)
+
+            distance_signed += distances[neighbor] * kernel_weight
+
+            normal_vector += normals[neighbor] * kernel_weight
+
+            volume += kernel_weight
+        end
+
+        if volume > eps()
+            distance_signed /= volume
+            normal_vector /= volume
+
+            # Store signed distance for visualization
+            system.signed_distances[particle] = distance_signed
+
+            constrain_particles!(u, system, particle, distance_signed, normal_vector)
+        end
+    end
+
+    return u
+end
+
+function constrain_particles!(u, system, particle, distance_signed, normal_vector)
+    (; shift_condition) = system
+
+    if distance_signed >= -shift_condition
+        # Constrain outside particles onto surface
+        shift = (distance_signed + shift_condition) * normal_vector
+
+        for dim in 1:ndims(system)
+            u[dim, particle] -= shift[dim]
+        end
+    end
+
+    system.is_boundary || return u
+
+    particle_spacing = system.initial_condition.particle_spacing
+    shift_condition_inner = system.tlsph ? particle_spacing : 0.5 * particle_spacing
+
+    if distance_signed < shift_condition_inner
+        shift = (distance_signed - shift_condition_inner) * normal_vector
+
+        for dim in 1:ndims(system)
+            u[dim, particle] -= shift[dim]
+        end
+    end
+
+    return u
+end
+
+@inline function update_transport_velocity!(system::ParticlePackingSystem, v_ode, semi)
+    v = wrap_v(v_ode, system, semi)
+    for particle in each_moving_particle(system)
+        for i in 1:ndims(system)
+            # Use the initial condition to store the evolved velocity
+            system.initial_condition.velocity[i, particle] = v[i, particle]
```

Comment:
Why? This is not GPU-compatible. The initial condition lies on the CPU. IMO, it would make more sense to store this in a dedicated vector. Also, the usual argument holds that you cannot call `solve` twice on this system and get the same results. You always have to re-create the system, which is annoying and confusing when working interactively.